netann/chan_status_manager.go

```diff
@@ -0,0 +1,595 @@
+			// if the inactive chan timeout has elapsed.
+			m.disableInactiveChannels()
+
+		case <-m.quit:
+			return
+		}
+	}
+}
+
+// processEnableRequest attempts to enable the given outpoint. If the method
+// returns nil, the status of the channel in chanStates will be
+// ChanStatusEnabled. If the channel is not active at the time of the request,
+// ErrEnableInactiveChan will be returned. An update will be broadcast only if
+// the channel is currently disabled, otherwise no update will be sent on the
+// network.
+func (m *ChanStatusManager) processEnableRequest(outpoint wire.OutPoint) error {
+	curState, err := m.getOrInitChanStatus(outpoint)
+	if err != nil {
+		return err
+	}
+
+	// Quickly check to see if the requested channel is active within the
+	// htlcswitch and return an error if it isn't.
+	chanID := lnwire.NewChanIDFromOutPoint(&outpoint)
+	if !m.cfg.IsChannelActive(chanID) {
+		return ErrEnableInactiveChan
+	}
+
+	switch curState.Status {
+
+	// Channel is already enabled, nothing to do.
+	case ChanStatusEnabled:
+		return nil
+
+	// The channel is enabled, though we are now canceling the scheduled
+	// disable.
+	case ChanStatusPendingDisabled:
+		log.Debugf("Channel(%v) already enabled, canceling scheduled "+
+			"disable", outpoint)
+
+	// We'll sign a new update if the channel is still disabled.
+	case ChanStatusDisabled:
+		log.Infof("Announcing channel(%v) enabled", outpoint)
+
+		err := m.signAndSendNextUpdate(outpoint, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	m.chanStates.markEnabled(outpoint)
+
+	return nil
+}
+
+// processDisableRequest attempts to disable the given outpoint. If the method
+// returns nil, the status of the channel in chanStates will be
+// ChanStatusDisabled. An update will only be sent if the channel has a status
+// other than ChanStatusEnabled, otherwise no update will be sent on the
+// network.
+func (m *ChanStatusManager) processDisableRequest(outpoint wire.OutPoint) error {
+	curState, err := m.getOrInitChanStatus(outpoint)
+	if err != nil {
+		return err
+	}
+
+	switch curState.Status {
+
+	// Channel is already disabled, nothing to do.
+	case ChanStatusDisabled:
+		return nil
+
+	// We'll sign a new update disabling the channel if the current status
+	// is enabled or pending-inactive.
+	case ChanStatusEnabled, ChanStatusPendingDisabled:
+		log.Infof("Announcing channel(%v) disabled [requested]",
+			outpoint)
+
+		err := m.signAndSendNextUpdate(outpoint, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	// If the disable was requested via the manager's public interface, we
+	// will remove the output from our map of channel states. Typically this
+	// signals that the channel is being closed, so this frees up the space
+	// in the map. If for some reason the channel isn't closed, the state
+	// will be repopulated on subsequent calls to RequestEnable or
+	// RequestDisable via a db lookup, or on startup.
+	delete(m.chanStates, outpoint)
```

Comment:
👍 

I think I ran into this issue in a prior diff of this PR. It appears to be fixed now.